File: app/src/main/java/me/sdk/GsonParser.java
package me.sdk;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import org.jetbrains.annotations.NotNull;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class GsonParser {

    public static JsonObject parse(@NotNull final String string) {
        try {
            final Gson gson = new Gson();
            return gson.fromJson(string, JsonObject.class);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static JsonObject parse(@NotNull final InputStream inputStream) {
        final String string = FileHelper.readString(inputStream);
        if (StringHelper.nullOrEmpty(string)) {
            LogHelper.warning("String was NULL");
            return null;
        }

        return parse(string);
    }

    public static List<String> keys(@NotNull final JsonObject jsonObject) {
        final List<String> keys = new ArrayList<String>();
        try {
            for (final Map.Entry<String, JsonElement> entry : jsonObject.entrySet()) {
                keys.add(entry.getKey());
            }
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
        }
        return keys;
    }

    public static List<JsonElement> values(@NotNull final JsonObject jsonObject) {
        final List<JsonElement> values = new ArrayList<JsonElement>();
        try {
            for (final Map.Entry<String, JsonElement> entry : jsonObject.entrySet()) {
                values.add(entry.getValue());
            }
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
        }
        return values;
    }

    public static boolean has(@NotNull final JsonObject jsonObject, @NotNull final String key) {
        try {
            return jsonObject.has(key);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return false;
        }
    }

    public static JsonObject object(@NotNull final JsonObject jsonObject, @NotNull final String key) {
        try {
            return jsonObject.getAsJsonObject(key);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static JsonArray array(@NotNull final JsonObject jsonObject, @NotNull final String key) {
        try {
            return jsonObject.getAsJsonArray(key);
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static String string(@NotNull final JsonObject jsonObject, @NotNull final String key) {
        try {
            return jsonObject.get(key).getAsString();
        }
        catch (final Exception e) {
            LogHelper.wtf(e);
            return null;
        }
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/datax/GsonParser.kt
package me.shkschneider.skeleton.datax

import com.google.gson.*
import me.shkschneider.skeleton.helperx.Logger
import java.util.*

object GsonParser : IParser<JsonObject, JsonArray> {

    override fun parse(string: String): JsonObject? {
        return parse(string, Gson())
    }

    fun parse(string: String, gson: Gson): JsonObject? {
        try {
            return gson.fromJson(string, JsonObject::class.java)
        } catch (e: JsonParseException) {
            Logger.wtf(e)
            return null
        }
    }

    override fun keys(jsonObject: JsonObject): List<String> {
        val keys = ArrayList<String>()
        for ((key) in jsonObject.entrySet()) {
            keys.add(key)
        }
        return keys
    }

    override fun values(jsonObject: JsonObject): List<JsonElement> {
        return jsonObject.entrySet().map { entry -> entry.value }
    }

    override fun has(jsonObject: JsonObject, key: String): Boolean {
        return jsonObject.has(key)
    }

    // from Object

    private fun element(jsonObject: JsonObject, key: String): JsonElement? {
        if (jsonObject.has(key)) {
            return jsonObject.get(key)
        }
        return null
    }

    override fun get(jsonObject: JsonObject, key: String, fallback: JsonObject?): JsonObject? {
        return element(jsonObject, key)?.asJsonObject
    }

    override fun array(jsonObject: JsonObject, key: String, fallback: JsonArray?): JsonArray? {
        return element(jsonObject, key)?.asJsonArray
    }

    override fun string(jsonObject: JsonObject, key: String, fallback: String?): String? {
        return element(jsonObject, key)?.asString ?: fallback
    }

    override fun number(jsonObject: JsonObject, key: String, fallback: Number?): Number? {
        return element(jsonObject, key)?.asNumber ?: fallback
    }

    override fun bool(jsonObject: JsonObject, key: String, fallback: Boolean?): Boolean? {
        return element(jsonObject, key)?.asBoolean ?: fallback
    }

    // from Array

    private fun element(jsonArray: JsonArray, index: Int): JsonElement? {
        return jsonArray.get(index) ?: null
    }

    override fun get(jsonArray: JsonArray, index: Int, fallback: JsonObject?): JsonObject? {
        return element(jsonArray, index)?.asJsonObject ?: fallback
    }

    override fun array(jsonArray: JsonArray, index: Int, fallback: JsonArray?): JsonArray? {
        return element(jsonArray, index)?.asJsonArray ?: fallback
    }

    override fun string(jsonArray: JsonArray, index: Int, fallback: String?): String? {
        return element(jsonArray, index)?.asString ?: fallback
    }

    override fun number(jsonArray: JsonArray, index: Int, fallback: Number?): Number? {
        return element(jsonArray, index)?.asNumber ?: fallback
    }

    override fun bool(jsonArray: JsonArray, index: Int, fallback: Boolean?): Boolean? {
        return element(jsonArray, index)?.asBoolean ?: fallback
    }

}
